NSStoryboard branch

gcasa opened the pull request:
I am working on implementing storyboards in GNUstep.  I am exposing this as a PR early on to get feedback during the entire process.  This addition will allow GNUstep to load storyboards and segues from .storyboard files.

fredkiefer commented on Headers/AppKit/NSSequePerforming.h:
```diff
@@ -0,0 +1,47 @@
+/* Definition of class NSSequePerforming
```
This class and the file and of course the implementation file as well should be called NSSeguePerforming. ("g" instead of "q")

gcasa replied:
Fixed.

fredkiefer commented on Headers/AppKit/NSStoryboard.h:
```diff
@@ -0,0 +1,67 @@
+
+@interface NSStoryboard : NSObject
+
++ (NSStoryboard *) mainStoryboard; // 10.13
```
If methods are only available for specific MacOS versions please surround them by corresponding #if statements.

fredkiefer replied:
Why did you mark this as resolved?

gcasa replied:
Sorry, it was premature.  I removed the ones in the .m file, but neglected the .h file.   It should be corrected now.

fredkiefer commented on Headers/AppKit/NSStoryboardSegue.h:
```diff
@@ -0,0 +1,47 @@
+#endif
+
+@interface NSStoryboardSegue : NSObject
+
```
Everything is still missing here.

gcasa replied:
Not for long. ;)

fredkiefer commented on Source/Functions.m:
```diff
@@ -44,6 +44,7 @@
 #import "AppKit/NSView.h"
 #import "AppKit/NSWindow.h"
 #import "AppKit/DPSOperators.h"
+#import "AppKit/NSStoryboard.h"
```
I do not see why you would need this import.

fredkiefer commented on Source/GSStoryboardLoader.m:
```diff
@@ -0,0 +1,218 @@
+/* <title>GSXibLoader</title>
```
Wrong file name and wrong attribution.

gcasa replied:
This class will be going away.  I don't need it and the other GSStoryboard classes.

fredkiefer commented on Source/GSStoryboardLoader.m:
```diff
@@ -0,0 +1,218 @@
+
+@implementation NSMenu (StoryboardCompatibility)
+
+- (BOOL) _isMainMenu
```
Now we have this method multiple times. Better move it to NSMenu.m and have it just once,

gcasa replied:
See above.

gcasa replied:
Class removed.

fredkiefer commented on Source/GSStoryboardLoader.m:
```diff
@@ -0,0 +1,218 @@
+}
+@end
+
+@implementation GSStoryboardLoader
```
The whole code in this class, apart from the first two methods looks identical to the class you copied it from. If this stays the case, please just subclass the original class.

gcasa replied:
Removed.

fredkiefer commented on Source/GSXib5KeyedUnarchiver.m:
```diff
@@ -170,6 +169,24 @@ @implementation GSXib5KeyedUnarchiver
 
+@interface NSString (XibExtension)
+
+- (BOOL) hasPrefixedClassName;
```
We should not in our library clutter up the namespace of classes from Foundation if this could be avoided easily.

gcasa replied:
Agreed, I'll do this another way.  I believe it can be undone at this point.

gcasa replied:
I may take this out if it's not needed.

gcasa replied:
Removed.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,63 @@
+   Boston, MA 02110 USA.
+*/
+
+#import <AppKit/NSStoryboard.h>
```
We discussed this already a hundred time, inside of gui you should be using "" for the AppKit imports.

gcasa replied:
Fixed.   This is an error in the template, I will correct it there as well.

fredkiefer commented on Source/GNUmakefile:
```diff
@@ -284,6 +286,9 @@ GSDisplayServer.m \
 GSHelpManagerPanel.m \
 GSInfoPanel.m \
 GSMemoryPanel.m \
+GSStoryboardDocument.m \
```
These classes are not part of the branch at the moment.

gcasa replied:
Removed them from my local and from the GNUmakefile they aren't needed.  I was considering another approach, but the current way seems better.   Storyboards are simply packed XIB files, so I am transforming them into proper XIBs before loading.

fredkiefer commented on Source/GSXib5KeyedUnarchiver.m:
```diff
@@ -207,6 +225,8 @@ + (void) initialize
                             @"NSWindowTemplate", @"window",
                             @"NSView", @"tableCellView",
                             @"IBUserDefinedRuntimeAttribute5", @"userDefinedRuntimeAttribute",
+                            @"GSStoryboardScene", @"scene",
```
Why would we want the create GNUstep specific objects here?????

gcasa replied:
There don't need to be any modifications to the XIB loading mechanism at this point, I am going to take these out as well as the modifications to handle the GS prefix.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,237 @@
+
+  if (success)
+    {
+      xml = [_scenesMap objectForKey: _applicationSceneId];
```
Why do you have the identical code twice?

gcasa replied:
This is a work in progress, one loads the application scene and the other loads the initial controller.  I am not sure if this should be split up or not.

gcasa replied:
I have now changed these to be distinct.

fredkiefer commented on Headers/AppKit/NSStoryboard.h:
```diff
@@ -0,0 +1,72 @@
+extern "C" {
+#endif
+
+  @class NSString, NSBundle, NSXMLDocument, NSArray, NSMutableDictionary;
```
Indentation is wrong and none of the additional classes you added in the last commit is really needed here.

gcasa replied:
Minimized and corrected.

fredkiefer commented on Source/GSXibLoader.m:
```diff
@@ -142,7 +140,7 @@ - (void) awake: (NSArray *)rootObjects
 - (BOOL) loadModelData: (NSData *)data
      externalNameTable: (NSDictionary *)context
-              withZone: (NSZone *)zone;
+              withZone: (NSZone *)zone
```
Good catch!

gcasa replied:
Thanks. :)

gcasa commented:
Fred, this is a WIP, I submitted it as a PR because I wanted your input on the method I am using to load the storyboards.  It's not ready for merging yet.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,237 @@
+      NSXMLDocument *storyboardXml = [[NSXMLDocument alloc] initWithData: data
+                                                                 options: 0
+                                                                   error: NULL];
+      AUTORELEASE(storyboardXml);
```
Why not move this after the next line and make it a RELEASE?

gcasa replied:
DONE

fredkiefer commented on Headers/AppKit/NSSeguePerforming.h:
```diff
@@ -0,0 +1,47 @@
+/* Definition of class NSSequePerforming
```
You seem to have renamed the file but not the class in it and all the other places where the wrong name gets used.

gcasa replied:
Ah, okay, sorry about that. :)

gcasa commented:
From GSXibLoader.m:

```
`  // Use the owner as first root object
  [(NSCustomObject*)[rootObjects objectAtIndex: 0] setRealObject: owner];`
```

That line is most certainly wrong.  I tested it with just a XIB file and it just so happens that an NSCustomObject5 is present as the first element, but it could be an NSWindow just as easily or an NSMenu.  I am going to make a change to this, but it will involve exposing "userLabel" in NSCustomObject.  The only other alternative is to test which one is an NSApplication or subclass thereof.

fredkiefer commented:
> From GSXibLoader.m:
> 
> ```
> `  // Use the owner as first root object
>   [(NSCustomObject*)[rootObjects objectAtIndex: 0] setRealObject: owner];`
> ```
> 
> That line is most certainly wrong. I tested it with just a XIB file and it just so happens that an NSCustomObject5 is present as the first element, but it could be an NSWindow just as easily or an NSMenu. I am going to make a change to this, but it will involve exposing "userLabel" in NSCustomObject. The only other alternative is to test which one is an NSApplication or subclass thereof.
> 
> Also the NSCustomObject5 which is presented in the rootObjects array is the wrong one.

Could you please provide me with a XIB file that shows this issue? I would try to find a proper solution for this. If this only happens for your converted storyboards a better solution would be to ensure the order of top level objects there.

gcasa commented:
> > From GSXibLoader.m:
> > ```
> > `  // Use the owner as first root object
> >   [(NSCustomObject*)[rootObjects objectAtIndex: 0] setRealObject: owner];`
> > ```
> > 
> > 
> > That line is most certainly wrong. I tested it with just a XIB file and it just so happens that an NSCustomObject5 is present as the first element, but it could be an NSWindow just as easily or an NSMenu. I am going to make a change to this, but it will involve exposing "userLabel" in NSCustomObject. The only other alternative is to test which one is an NSApplication or subclass thereof.
> > Also the NSCustomObject5 which is presented in the rootObjects array is the wrong one.
> 
> Could you please provide me with a XIB file that shows this issue? I would try to find a proper solution for this. If this only happens for your converted storyboards a better solution would be to ensure the order of top level objects there.


Sure...  this actually happens in almost EVERY SINGLE XIB...   The XIB I have attached is for NSSwitch_test (you can find this project in my gcasa account).   The XIB loads and it does work, but the first NSCustomObject in the list is FirstResponder NOT File's Owner, but because the XIB loading code ASSUMES that the first in the list will be File's Owner it just stuffs it in.  Also, later in the code it skips 3 objects presuming that it is File's Owner, First Responder, and NSApplication and puts the rest into the topLevelObjects dictionary.    In my opinion it is much better to look for the right place to put it rather than assume any kind of order.

[MainMenu.xib.zip](https://github.com/gnustep/libs-gui/files/4821198/MainMenu.xib.zip)

If you absolutely disagree with this I will change it back and find another way to make it work.

fredkiefer commented:
> > > From GSXibLoader.m:
> > > ```
> > > `  // Use the owner as first root object
> > >   [(NSCustomObject*)[rootObjects objectAtIndex: 0] setRealObject: owner];`
> > > ```
> > > That line is most certainly wrong. I tested it with just a XIB file and it just so happens that an NSCustomObject5 is present as the first element, but it could be an NSWindow just as easily or an NSMenu. I am going to make a change to this, but it will involve exposing "userLabel" in NSCustomObject. The only other alternative is to test which one is an NSApplication or subclass thereof.
> > > Also the NSCustomObject5 which is presented in the rootObjects array is the wrong one.
> > 
> > Could you please provide me with a XIB file that shows this issue? I would try to find a proper solution for this. If this only happens for your converted storyboards a better solution would be to ensure the order of top level objects there.
> 
> Sure... this actually happens in almost EVERY SINGLE XIB... The XIB I have attached is for NSSwitch_test (you can find this project in my gcasa account). The XIB loads and it does work, but the first NSCustomObject in the list is FirstResponder NOT File's Owner, but because the XIB loading code ASSUMES that the first in the list will be File's Owner it just stuffs it in. Also, later in the code it skips 3 objects presuming that it is File's Owner, First Responder, and NSApplication and puts the rest into the topLevelObjects dictionary. In my opinion it is much better to look for the right place to put it rather than assume any kind of order.
> 
> [MainMenu.xib.zip](https://github.com/gnustep/libs-gui/files/4821198/MainMenu.xib.zip)
> 
> If you absolutely disagree with this I will change it back and find another way to make it work.


The argument was whether the file owner was the first custom object in the list or not. And you wanted to provide a file where this wasn't the case. Here is an extract from your file:

    <objects>
        <customObject id="-2" userLabel="File's Owner" customClass="NSApplication">
            <connections>
                <outlet property="delegate" destination="Voe-Tx-rLC" id="GzC-gU-4Uq"/>
            </connections>
        </customObject>
        <customObject id="-1" userLabel="First Responder" customClass="FirstResponder"/>
        <customObject id="-3" userLabel="Application" customClass="NSObject"/>

As you can see the file owner is the first object here. It could still be that the objects are processed in a different order by our code. In that case we should analyse why. But first could you please validate that this issue actually exists on the file you attached? I don't want to waste my time chasing after a non existing issue.

fredkiefer commented:
> As you can see the file owner is the first object here. It could still be that the objects are processed in a different order by our code. In that case we should analyse why. But first could you please validate that this issue actually exists on the file you attached? I don't want to waste my time chasing after a non existing issue.

As you did not reply, I wasted my time and checked the behaviour by adding these lines

  NSLog(@"First object %@", [rootObjects objectAtIndex: 0]);
  NSLog(@"Second object %@", [rootObjects objectAtIndex: 1]);
  NSLog(@"Third object %@", [rootObjects objectAtIndex: 2]);

Right before the line that sets the file owner and this is the output when loading your XIB file:

2020-06-27 22:18:30.324 NSSwitch_test[24881:24881] First object <NSCustomObject5: 2710320> = <<className: NSApplication, object: (null)>>
2020-06-27 22:18:30.324 NSSwitch_test[24881:24881] Second object <NSCustomObject5: 2920400> = <<className: FirstResponder, object: (null)>>
2020-06-27 22:18:30.324 NSSwitch_test[24881:24881] Third object <NSCustomObject5: 296b910> = <<className: NSApplication, object: (null)>>

This may look a bit confusing as we do not output the user label and the decoding code for NSCustomObject5 changes the className for the object with the label Application to NSApplication. But if you take that into account, thinks look completely correct to me. 
Sorry, as long as these changes are in I will have to reject this pull request.

gcasa commented:
I didn't ask you to do so yet as I still need to handle segues.  Also, when
I tested it the object it was setting it into was NOT the file's owner.  I
believe it's wrong to assume it's always first.   I will see what I can do
with the XML manipulation to make sure the object is first so that the
previous code will work.  I will further verify (and maybe convince myself)
that it will work reliably.

gcasa commented:
Sorry about not replying... I do have other work sometimes.

gcasa commented:
> > > > From GSXibLoader.m:
> > > > ```
> > > > `  // Use the owner as first root object
> > > >   [(NSCustomObject*)[rootObjects objectAtIndex: 0] setRealObject: owner];`
> > > > ```
> > > > 
> > > > 
> > > > That line is most certainly wrong. I tested it with just a XIB file and it just so happens that an NSCustomObject5 is present as the first element, but it could be an NSWindow just as easily or an NSMenu. I am going to make a change to this, but it will involve exposing "userLabel" in NSCustomObject. The only other alternative is to test which one is an NSApplication or subclass thereof.
> > > > Also the NSCustomObject5 which is presented in the rootObjects array is the wrong one.
> > > 
> > > 
> > > Could you please provide me with a XIB file that shows this issue? I would try to find a proper solution for this. If this only happens for your converted storyboards a better solution would be to ensure the order of top level objects there.
> > 
> > 
> > Sure... this actually happens in almost EVERY SINGLE XIB... The XIB I have attached is for NSSwitch_test (you can find this project in my gcasa account). The XIB loads and it does work, but the first NSCustomObject in the list is FirstResponder NOT File's Owner, but because the XIB loading code ASSUMES that the first in the list will be File's Owner it just stuffs it in. Also, later in the code it skips 3 objects presuming that it is File's Owner, First Responder, and NSApplication and puts the rest into the topLevelObjects dictionary. In my opinion it is much better to look for the right place to put it rather than assume any kind of order.
> > [MainMenu.xib.zip](https://github.com/gnustep/libs-gui/files/4821198/MainMenu.xib.zip)
> > If you absolutely disagree with this I will change it back and find another way to make it work.
> 
> The argument was whether the file owner was the first custom object in the list or not. And you wanted to provide a file where this wasn't the case. Here is an extract from your file:
> 
> ```
> <objects>
>     <customObject id="-2" userLabel="File's Owner" customClass="NSApplication">
>         <connections>
>             <outlet property="delegate" destination="Voe-Tx-rLC" id="GzC-gU-4Uq"/>
>         </connections>
>     </customObject>
>     <customObject id="-1" userLabel="First Responder" customClass="FirstResponder"/>
>     <customObject id="-3" userLabel="Application" customClass="NSObject"/>
> ```
> 
> As you can see the file owner is the first object here. It could still be that the objects are processed in a different order by our code. In that case we should analyse why. But first could you please validate that this issue actually exists on the file you attached? I don't want to waste my time chasing after a non existing issue.

You are right.  I was confused when I saw "Application" or "NSApplication" during debugging and thought things were out of order.  I have changed the code in NSStoryboard to put the customObjects in the correct order.  I have reverted the code in GSXibLoader.m to what it was previously.  Sorry about going down a blind alley on this one.  It's actually better now that it is compliant.

gcasa commented:
Fred, please review this. Currently segues work and loading of external storyboards are working.  This is a large and complex change.  I am planning to do some refactoring and there are some missing features (such as the fact that NSWindowController and NSViewController do not unarchive certain attributes).   I think much of this is best handled in a separate branch as the bulk of the functionality is implemented here.  I am also planning on moving much of the transformation logic into its own class once I get a handle on how to best do that.

fredkiefer commented on Source/Functions.m:
```diff
@@ -85,6 +94,29 @@
+        }
+      else
+        {
+          NSLog(_(@"Could not find model to load %@"), mainModelFile);
```
Shouldn't this error message talk about story board and in the current code the displayed value will always either be nil or empty.

fredkiefer commented on Source/GSXib5KeyedUnarchiver.h:
```diff
@@ -28,6 +28,8 @@
 
 #import <Foundation/Foundation.h>
 #import "GNUstepGUI/GSXibKeyedUnarchiver.h"
+#import "GNUstepGUI/GSNibLoading.h"
```
Please remove these two lines as they are not needed.

fredkiefer commented on Source/GSXib5KeyedUnarchiver.m:
```diff
@@ -449,7 +449,8 @@ - (NSString*) alternateName: (NSString*)name
   NSString *typeName = [currentElement attributeForKey: @"key"];
 
-  if ((typeName != nil) && [postfix hasPrefix: typeName])
+  if ((typeName != nil) && [postfix hasPrefix: typeName] &&
```
Why is this part of your pull request? This is already on the master branch.

gcasa replied:
I had backed out some changes on my branch.  I did so by copying the file from master, I'll make sure this doesn't accidently back anything out on master.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+/* Implementation of class NSStoryboard
+   Copyright (C) 2019 Free Software Foundation, Inc.
```
Wrong year

gcasa replied:
fixed

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+        {
+          [self setKind: [coder decodeObjectForKey: @"NSKind"]];
+        }
+    }
```
The else branch is missing.

gcasa replied:
Intentionally, it's being unarchived from the XML only.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+  // this is never encoded directly...
```
Code is missing

gcasa replied:
Intentionally, it will never be encoded.  This is being decoded from the transformed XML *only*

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+  self = [super init];
+  if (self != nil)
+    {
+      _storyboardName = nil;
```
This is the default in Objective C, no code is needed for that.

gcasa replied:
Removed

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+- (id) copyWithZone: (NSZone *)z
+{
+  NSControllerPlaceholder *c = [[NSControllerPlaceholder allocWithZone: z] init];
+  [c setStoryboardName: _storyboardName];
```
What about the storyboard?

gcasa replied:
Removed and made this a local var in `- (id) instantiate`

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+        {
+          [self setStoryboardName: [coder decodeObjectForKey: @"NSStoryboardName"]];
+        }
+    }
```
The else branch is missing

gcasa replied:
Intentionally as above.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+  // this is never encoded directly...
```
Code is missing

gcasa replied:
Intentionally as above.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+
+- (id) instantiate
+{
+  _storyboard = [NSStoryboard storyboardWithName: _storyboardName
```
Could this become a local variable as it is never used as an instance variable?

gcasa replied:
Done

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+      
+      while ((e = [en nextObject]) != nil)
+        {
+          NSXMLElement *doc = [[NSXMLElement alloc] initWithName: @"document"];
```
This method is way too long. Could you at least move the code in this loop into a helper method?

gcasa replied:
This is part of my planned refactoring if we can defer this. I would appreciate it.  I am planning on moving all of the transformation logic out of this file.

gcasa replied:
Deferred until refactor.

fredkiefer replied:
IN software engineering we call this the "big refactor in the sky" or "grand rewrite in the sky". Feel free to check this out.

gcasa replied:
LOL, true.  It's been done.  I am a Computer Scientist, so it's cool. ;)

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+          NSString *controllerId = nil;
+
+          // Copy children...
+          while ((child = [ce nextObject]) != nil)
```
Why don't you use the FOR_IN fast enumeration code here and in the outer loop?

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+              
+              NSArray *subnodes = [child nodesForXPath: @"//application" error: NULL];
+              NSXMLNode *appNode = [subnodes objectAtIndex: 0];
+              if ([[appNode name] isEqualToString: @"application"] == YES)
```
Isn't this already enforced by the XPath above?

gcasa replied:
Fixed

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+              NSXMLNode *appNode = [subnodes objectAtIndex: 0];
+              if ([[appNode name] isEqualToString: @"application"] == YES)
+                {
+                  NSXMLElement *objects = (NSXMLElement *)[appNode parent];
```
Shouldn't we use a better name here, for example toplevelobjects?

gcasa replied:
objects is the <objects> section in the XML.   That's why I am calling it that. :)

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+  return customObject;
+}
+
+- (void) _processStoryboard: (NSXMLDocument *)storyboardXml
```
Wouldn't this whole method be better expressed as an XSLT?

gcasa replied:
Most likely, but it's done now and it is working perfectly.  It would be interesting to write a transform for it, but I will save that for the refactor.

gcasa replied:
Deferred for refactoring.  I think doing this via XSLT might make it much more compact.  Doing it in ObjC code helped me understand everything which needs to be done.  What would concern me is making XSLT a hard dependency to support this.

gcasa replied:
Done

gcasa replied:
I really don't think we are ever going to re-write this in XSLT.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+                }
+              else
+                {
+                  NSXMLElement *customObject = nil;
```
Most of the code in this branch is very similar to the if branch, wouldn't it be possible to share more of the code?

gcasa replied:
I believe so.  I will try to do this now.

gcasa replied:
I'll save this for the refactor as well.  Since I am planning on reducing and moving all of the code.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+                             forKey: sceneId];
+              
+              // Map controllerId's to scenes...
+              if (controllerId != nil)
```
Does this mean there will ever by only one controller in any story board file? The above code suggests that this should rather be called for every controller we find in the file.

gcasa replied:
There will actually only be one controller in each scene.   I tried multiple ways to prove this to myself when working in Xcode.  Additionally, if you look at the "compiled" storyboard which, basically ends up being a bunch of nibs, there is always one controller in each scene.   the purpose of this variable is to make each controllerId to it's corresponding scene so that we can instantiate the proper objects when instantiateWithControllerId: is called.

fredkiefer replied:
If this is the case, the code above should express this better. Maybe have a Sparte method for each  scene transformation and another method to get the one controller for each scene?

gcasa replied:
Done.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+  NSMapTable *mapTable = [NSMapTable strongToWeakObjectsMapTable];
+  NSArray *connectionsArray = [xmlIn nodesForXPath: @"//connections"
+                                             error: NULL];
+  NSArray *array = [xmlIn nodesForXPath: @"//objects"
```
Couldn't you use @"//objects[1]" here?

gcasa replied:
Done.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+  NSArray *array = [xmlIn nodesForXPath: @"//objects"
+                                  error: NULL];
+  NSXMLElement *objects = [array objectAtIndex: 0]; // get the "objects" section
+  NSArray *controllers = [objects nodesForXPath: @"windowController"
```
Shouldn't this be moved down where the value gets used for the first time?

gcasa replied:
Done.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+        }
+    }
+  
+  NSString *src = nil;
```
Declaring a variable in the middle of a method will create issues with old compilers or compilers in old mode. I am all for allowing this but we should make sure this is nonsense amongst GNUstep developers.

gcasa replied:
Do we support any compilers which disallow this?

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+    }
+}
+
+- (NSMapTable *) _processSegues: (NSXMLDocument *)xmlIn
```
Again a method that would benefit from being split up.

gcasa replied:
Yes, indeed, both methods have gotten a bit big.   Again, I want to pull everything into another class and, possibly, do the XSLT change you suggested.

gcasa replied:
While it could be done using XSLT, I believe doing it that way would defeat the idea of maintainability.  Also, I don't want to make XSLT a hard dependency for supporting this.   Additionally, I believe that doing it in ObjC is faster, though I have no tests to prove this.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+  self = [super init];
+  if (self != nil)
+    {
+      NSString *path = [bundle pathForResource: name
```
Should we use a local autorelease pool here?

gcasa replied:
Not sure.

gcasa replied:
At this point I don't think so.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+    }
+}
+
++ (NSStoryboard *) mainStoryboard // 10.13
```
This comment is not needed here

fredkiefer commented on Source/NSStoryboardSegue.m:
```diff
@@ -0,0 +1,153 @@
+  _sourceController = controller;
+}
+
++ (instancetype)segueWithIdentifier: (NSStoryboardSegueIdentifier)identifier 
```
Sometimes you have a space after the return type and sometimes you don't. What is the point here?

fredkiefer commented on Source/NSStoryboardSegue.m:
```diff
@@ -0,0 +1,153 @@
+      else
+        {
+          NSWindow *w = [NSWindow windowWithContentViewController: _destinationController];
+          [w setTitle: [_destinationController title]];
```
Why don't we need this call in the "modal" case?

fredkiefer commented on Source/NSViewController.m:
```diff
@@ -179,6 +190,48 @@ - (void) encodeWithCoder: (NSCoder *)aCoder
+      else
+        {
+          NSString *destId = [segue destinationController];
+          destCon = [_storyboard instantiateControllerWithIdentifier: destId]; 
```
Where is _storyboard ever set?

gcasa replied:
`      
      while ((o = [en nextObject]) != nil)
        {
          if ([o isKindOfClass: [NSWindowController class]])
            {
              wc = (NSWindowController *)o;
              [wc _setSegueMap: segueMap];
              [wc _setTopLevelObjects: topLevelObjects];
              [wc _setStoryboard: self];
              [wc _setOwner: NSApp];
              result = o;`

gcasa replied:
NSStoryboard.m in the _processStoryboard: method.

fredkiefer replied:
My feeling here is that most of this code actually belongs into the segue. But to be sure I should be able to better understand all your transformation code to see whether stuff could be easily moved around. Why doesn't the segue know about the owning story board and and computes it destination controller all by itself?

gcasa replied:
The only time the code can get the destination controller is when the segue is initiated.  Until then it's necessary to simply store the id.

gcasa replied:
I don't think this is needed at this point. ;)

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+}
+
+- (id) instantiateControllerWithIdentifier: (NSStoryboardSceneIdentifier)identifier
+                                   creator: (NSStoryboardControllerCreator)block  // 10.15
```
Comment is not needed

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,975 @@
+
+- (id) instantiateInitialControllerWithCreator: (NSStoryboardControllerCreator)block // 10.15
+{
+  id controller = [self instantiateInitialController];
```
Normally we call the more general method from the specific one, here you do it the other way around

gcasa replied:
Fixing this now.

gcasa replied:
Done.

fredkiefer commented on Headers/AppKit/NSStoryboard.h:
```diff
@@ -0,0 +1,79 @@
+- (id) instantiateInitialController;
+
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_15, GS_API_LATEST)
+- (id) instantiateInitialControllerWithCreator: (NSStoryboardControllerCreator)block; // 10.15
```
The comment here could now be removed, as could the one further down.

gcasa replied:
Removed.

fredkiefer commented on Source/NSWindow.m:
```diff
@@ -723,6 +724,24 @@ + (void) initialize
+                                                  defer: NO];
+  [window setTitle: title];
+  [window setContentView: view];
+  return window;
```
There seems to be an autorelease missing here.

gcasa replied:
done.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,967 @@
+
+@implementation NSControllerPlaceholder
+
+- (instancetype) init
```
With just that code it is better to completely remove this method.

gcasa replied:
gone

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,967 @@
+      _controllerMap = [[NSMutableDictionary alloc] initWithCapacity: [array count]];
+      _documentsMap = [[NSMutableDictionary alloc] initWithCapacity: [array count]];
+      
+      //while ((e = [en nextObject]) != nil)
```
Left over code fragment.

gcasa replied:
Gone

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,967 @@
+          NSString *sceneId = [[e attributeForName: @"sceneID"] stringValue]; 
+          NSString *controllerId = nil;
+
+          // Copy children...
```
Actually you are not coming these, just moving.

gcasa replied:
You're right, changed

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,967 @@
+                  [doc addChild: objects];
+                  
+                  // Assign application scene...
+                  ASSIGN(_applicationSceneId, sceneId);
```
Why is this line here and not up there where we get the value?

gcasa replied:
Added it to where we know we are in the application scene, but after declarations.

gcasa commented:
Added an issue for the refactoring part.

gcasa commented:
I have refactored everything which pertains to the transformation into a separate class.

fredkiefer commented on Headers/AppKit/NSStoryboard.h:
```diff
@@ -0,0 +1,75 @@
+  id _transform;
+}
+  
++ (NSStoryboard *) mainStoryboard;
```
This is a 10.13 method and you previously had a comment about that, what is the problem with surrounding this method with 
#if OS_API_VERSION(MAC_OS_X_VERSION_10_13, GS_API_LATEST)

gcasa replied:
I didn't realize it was.  Fixed now.

fredkiefer commented on Source/Functions.m:
```diff
@@ -85,6 +94,29 @@
+            }
+          else
+            {
+              [NSStoryboard _setMainStoryboard: storyboard];
```
These three methods always get called together. Wouldn't it make more sense to have just one method here and have the rest over at NSStoryBoard?

gcasa replied:
There is no documented method which does all three, so it would need to be private.  That's the only reason I didn't do it.

fredkiefer replied:
You already use two undocumented methods here. This does not sound like a real argument.

gcasa replied:
This code occurs once, so I don't think it's really necessary to do this now.

fredkiefer replied:
The idea here was to have just the method _instantiateApplicationScene and in the implementation of that method call _setMainStoryboard:  and instantiateInitialController. Or if you would prefer to stick with the one public method have instantiateInitialController call the other two methods. The net result is always the sam, but we have less inter dependency and that by itself is a very good thing.

fredkiefer replied:
OK, I just checked. The last option won't work because the method instantiateInitialController is also used for placeholders. So the best way is to hide everything behind _instantiateApplicationScene.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,747 @@
+  return _documentsMap;
+}
+
+- (NSDictionary *) identifierToSegueMap
```
Is this ever used?

gcasa replied:
Yes, it's used to get the segueMap to set on the controller in processSegues:forId: and subsequently the identifier is used to resolve the destination controller.

fredkiefer replied:
I was referring to this method and this indeed never is used. The next method fills the role you are referring to.

gcasa replied:
Old habit, I just wanted a way to get it if needed.  I will remove it.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,747 @@
+  return [xml XMLData];
+}
+
+- (NSData *) dataForSceneId: (NSString *)sceneId
```
Looking at the difference in code between this method and the one before, I am wondering whether we should store the application scene in a different place?

gcasa replied:
I have optimized this code somewhat.  There is now a special identifier for the "application" scene.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,747 @@
+              // Map controllerId's to scenes...
+              if (controllerId != nil)
+                {
+                  [_controllerMap setObject: sceneId
```
You could call processSegues:forId: here already and store the result of that in a map.

gcasa replied:
Code has been restructured.

fredkiefer replied:
All the restructuring is good but it does not address the point I am making. What I was suggesting was to do all the XML transformations here. That way we would already store the results (Maybe even in just one map not with two) and the code later would just read from that map.

gcasa replied:
The changes work.  They are efficient.  Why put all of the transformations in one place?

gcasa replied:
The changes work.  They are efficient.  Why put all of the transformations in one place?

fredkiefer replied:
Sorry, we seem to have different goals here. You seem to want code that is basically working. Working code is important but for me this is where it starts. What we need is code that is maintainable. The minimal requirement for that is that the complexity of the code does not add to the complexity of the problem it is trying to solve. Otherwise over time the existing code becomes the actual problem. At the moment you understand the problem and your code. This is the best time to improve the code to really express the problem. In a few months time nobody will remember why the code is that way it was written and to change it will then require a much bigger effort.

gcasa replied:
*pinches bridge of nose*  okay, fred.

gcasa replied:
I moved processSegues:... into the method which processes the storyboard.  The result is stored in the identifierToSegueMap  (this is the segue id).  My goal is the same as yours.  I prefer maintainable code, but I also like to take an iterative approach and optimize it over time.  I really think that's the only difference.   I, however, concede to your judgment as you're the GUI maintainer.  I won't merge without your approval.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,747 @@
+  return [self dataForSceneId: _applicationSceneId];
+}
+
+- (void) processStoryboard: (NSXMLDocument *)storyboardXml
```
Now you have a separate class for the transformation, but the internal structure isn't any better. You still need to split out helper methods and add a few more FOR_IN usages and move the code around a bit.

gcasa replied:
Yes, I know that.  Since everything is working and well tested, I would like to push to master and work on optimization as a separate task.

gcasa replied:
All code has been broken out into manageable methods and FOR_IN is used universally

gcasa commented:
Fred, please review.  I would like to merge this soon.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,276 @@
+            {
+              NSStoryboardSeguePerformAction *ssa = (NSStoryboardSeguePerformAction *)o;
+              [ssa setSender: result]; // resolve controller here...
+              [ssa setIdentifierToSegueMap: [_transform identifierToSegueMap]];
```
As far as I understand this map exists only once in a storyboard. Why not keep it on the storyboard itself instead of the action?

gcasa replied:
it's actually calculated per controller.   So, while it could exist on the storyboard, it's easier to resolve it if it's on the action.

fredkiefer replied:
If this was your intention then the actual code is wrong. There is only one such mach and with in the NSStoryboardSeguePerformAction you only ever use the key "application".
Not sure what to believe, the code or your explanation.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,866 @@
+    }
+  else // This is a special case where there is no source controller and we don't ask "should"
+    {
+      NSMapTable *mapTable = [_identifierToSegueMap objectForKey: @"application"];
```
I will try to explain in more detail the changes I am suggesting. (But I don't have the time at the moment to do so for each place, sorry.)
 For example here this is the only place where the NSStoryboardSeguePerformAction uses the _identifierToSegueMap. You could try to store the NSStoryboardSegue directly in the NSStoryboardSeguePerformAction in the method instantiateControllerWithIdentifier:creator:. That way the internal complexity of the code would be greatly reduced. 
Of course this is only possible if the object get created  in the correct order. But I think this is possible.

gcasa replied:
One of the reasons I did it the way it is currently done is to delay the processing of the scenes until they are needed.   I think I understand what you're saying, but I would ask that we defer this change to a later date.    The changes, as they are, work very well and are well tested.  I will take a look into what you're talking about.

gcasa replied:
I think this is something BETTER done in a later iteration.  The current code works.

gcasa replied:
I'm deferring this to a later commit.  I really want to be done with this code.

gcasa replied:
The issue with storing NSStoryboardSegue this way is that the NSStoryboardSegueAction isn't unarchived until it's loaded in instantiateControllerWithIdentifier: method in NSStoryboard.  This is why I had to organize the code in the way I did by mapping the segue itself to the Segue identifiers associated with a given controller.   So we have a map of controllers, then we have a map of segues which belong to each of those controllers.  When the controller is instantiated (based on it's sceneId) the XML I created for the NSStoryboardSegueAction is then unarchived.   

Either way we will need to unarchive the scene/document in order to story the segue on the action which is what we are pretty much doing anyway.

gcasa replied:
I will go ahead and make this change and see if it simplifies the code.  Since we have everything we need in the instantiateControllerWithIdentifier:... method.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,866 @@
+      [_identifierToSegueMap setObject: mapTable
+                                forKey: identifier];                  
+      // Add to cache...
+      [_documentsMap setObject: mapTable
```
I could not find any place where this is ever used. Could you please explain the idea?

gcasa replied:
I've removed it.  After the reorganization, it is not needed.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,866 @@
+      NSString *value = [a stringValue];
+      if (value != nil)
+        {
+          [uuid appendString: value];
```
You never use the uuid string in the case where this method return YES. Maybe you could turn the uuid into the return value and only return it when a new one gets created? Or could all of that be simplified even more?

gcasa replied:
Currently if this is set, then it is assumed that the document has been processed.  I suppose I could replace this with just a "true" or "false" value.

gcasa replied:
Done.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,866 @@
+}
+
+- (NSString *) findControllerIdInDocument: (NSXMLDocument *)xmlIn
+                              withObjects: (NSXMLElement *)objects
```
You should be using singular for this parameter.

gcasa replied:
It's passing in the "objects" section of the XML, that's why it's plural.

gcasa commented:
Hey Fred, please take a look again, I have made all of the requested changes.

gcasa commented:
Awaiting approval.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,280 @@
+      END_FOR_IN(seguesToPerform);
+
+      // process placeholders...
+      FOR_IN(id, o, placeholders)
```
I am not sure here, but wouldn't it simplify the logic in NSStoryboardSeguePerformAction if we instantiated the placeholders first?

gcasa replied:
Not greatly, but it's best to resolve them first before doing any segues.

fredkiefer replied:
Why do you use this intermediate list placeholders at all and not instantiate the placeholders in the first loop over topLevelObjects?

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,280 @@
+          if ([o isKindOfClass: [NSStoryboardSeguePerformAction class]])
+            {
+              NSStoryboardSeguePerformAction *ssa = (NSStoryboardSeguePerformAction *)o;
+              NSMapTable *mapTable = [[_transform identifierToSegueMap] objectForKey: identifier];
```
You should be using segueMapForIdentifier: here and then you could actually remove the method identifierToSegueMap as you already promised.

gcasa replied:
Done

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,879 @@
+  ASSIGN(_storyboardSegue, ss);
+}
+
+- (id) nibInstantiate
```
If this is the correct implementation then you don't need this method.

gcasa replied:
Removed

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,879 @@
+    }
+  else // This is a special case where there is no source controller and we don't ask "should"
+    {
+      id destCon = nil;
```
We could move all of the destinationController handling before the if statement on line 183. That way we could remove the code duplication in NSViewController and NSWindowController and only have the code once here.

There surely is an even better way to resolve this, but for now that move would be good enough.

gcasa replied:
Hmm we don't want to instantiate the destination controller unless the answer from "shouldPerform" on the controller is "YES" except in the special case.    Working on this.

gcasa replied:
Done.

gcasa commented:
Awaiting approval.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,878 @@
+        {
+          [_sender performSegueWithIdentifier: _identifier
+                                       sender: _sender];
+        }
```
The else case is missing here

gcasa replied:
Why? Should remains YES.

gcasa replied:
I added an explanation as to why there is no else.

fredkiefer replied:
I am not able to see that explanation and previously you had a [_storyboardSegue perform]; here. Was that code wrong?

fredkiefer replied:
As nil does not respond to any selector you could now remove that check. But more importantly you are still missing the [_storyboardSegue perform] call in the else case of this if statement.
Sometimes I have the impression that I am not expressing my comments clear enough. On the other hand I don't know what else to write to make it more explicit.

fredkiefer replied:
Do you actually read my comments? Sometimes I doubt it.

gcasa replied:
"an else case is missing" that's really explicit, Fred. </sarcasm>

fredkiefer replied:
OK, let's try again:

Could you please read the code 

else
{
 [_storyboardSegue perform]; 
}

here?

gcasa replied:
Given that the code below that is performing that function fine is why I was confused.  If you look at the code below it is resolving the destination controller which *still* needs to be looked up even in the special case.  So I was not seeing your point.    The additional check that is done on _sender is to make certain that the performSegueWithIdentifier: method is not called on an object that doesn't have it.  The else fits best there.   Sorry for the misunderstanding.

fredkiefer commented on Source/NSViewController.m:
```diff
@@ -29,11 +29,20 @@
+#import "AppKit/NSStoryboard.h"
+#import "AppKit/NSWindowController.h"
 
+@interface NSStoryboardSegue (__ViewControllerPrivate__)
```
These are no longer needed

gcasa replied:
Removed

fredkiefer commented on Source/NSWindowController.m:
```diff
@@ -33,13 +33,23 @@
+
 #import "NSDocumentFrameworkPrivate.h"
 
+@interface NSStoryboardSegue (__WindowControllerPrivate__)
```
These are no longer needed

gcasa replied:
Removed

fredkiefer commented on Source/GSStoryboardTransform.h:
```diff
@@ -0,0 +1,122 @@
+
+- (instancetype) initWithData: (NSData *)data;
+
+- (NSDictionary *) scenesMap;
```
This never gets used from the outside.

gcasa replied:
Removed.

fredkiefer commented on Source/GSStoryboardTransform.h:
```diff
@@ -0,0 +1,122 @@
+- (instancetype) initWithData: (NSData *)data;
+
+- (NSDictionary *) scenesMap;
+- (NSDictionary *) controllerMap;
```
Neither is this.

gcasa replied:
Removed.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,878 @@
+
+  if (responds)
+    {
+      should = [_sender shouldPerformSegueWithIdentifier: _identifier
```
Why do you treat this method as optional whereas the other two from the protocol are handled as being required?

gcasa replied:
In the NSStoryboard_test example, there Is a special case where a segue can be invoked from a non -controller (i.e. an NSMenuItem).  So, on those objects, that method won't be there.

gcasa replied:
I just now pushed a check for this.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,871 @@
+  // if it is not, as it is possible to initiate a segue from an NSMenuItem, then we don't, but should
+  // remains set to YES so that the logic to replace the destination controller is still called.
+  if ([_sender respondsToSelector: @selector(shouldPerformSegueWithIdentifier:sender:)]  &&
+      ([_sender isKindOfClass: [NSViewController class]] || [_sender isKindOfClass: [NSWindowController class]]))
```
I think these extra test here are not required. Rather you should add a test further down before calling performSegueWithIdentifier:sender:

gcasa replied:
What test specifically?

gcasa replied:
Added a check to see if the _sender responds to the performSegueForIdentifier:... method.

gcasa commented:
Awaiting... approval...

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,871 @@
+  NSString *customClassString = nil;
+  if (appNode != nil)
+    {
+      NSXMLElement *objects = (NSXMLElement *)[appNode parent];
```
The node "objects" will always be the same node as "child". The code here would be a lot easier to understand if you handed hin "child" with the name "objects" and used it consistently

gcasa replied:
I have refactored this to use your code.  There is a difference here.  When the <application> node is present, the <objects> node appears within it, so it's necessary to pull that out.  But it does reside in a different place.

fredkiefer replied:
Actually I was thinking more in the line of this structure:

`- (void) processChild: (NSXMLElement *)objects
               withDoc: (NSXMLElement *)doc
           withAppNode: (NSXMLNode *)appNode
               sceneId: (NSString *)sceneId
 {
   NSString *customClassString = nil;
   NSXMLNode *appCons = nil;

   if (appNode != nil)
     {
       NSArray *appConsArr = [appNode nodesForXPath: @"connections" error: NULL];

       appCons = [appConsArr objectAtIndex: 0];
       if (appCons != nil)
         {
           [appCons detach];
         }

       // Assign application scene...
       ASSIGN(_applicationSceneId, sceneId);
       [_controllerMap setObject: _applicationSceneId
                          forKey: APPLICATION];

       // Move all application children to objects...
       NSArray *appChildren = [appNode children];      
       FOR_IN(NSXMLElement*, ae, appChildren)
         {
           [ae detach];
           [objects addChild: ae];
         }
       END_FOR_IN(appChildren);

       // Remove the appNode
       [appNode detach];

       // create a customObject entry for NSApplication reference...
       NSXMLNode *appCustomClass = [(NSXMLElement *)appNode
                                                    attributeForName: @"customClass"];
       customClassString = ([appCustomClass stringValue] == nil) ?
         @"NSApplication" : [appCustomClass stringValue];
     }

     [self addStandardObjects: objects
                    classString: customClassString
                    connections: appCons];

     // Add it to the document
     [objects detach];
     [doc addChild: objects;
 }`

Removing a code duplication and an unneeded type cast along the way.

gcasa replied:
This is done.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,871 @@
+                                                   attributeForName: @"customClass"];
+      customClassString = ([appCustomClass stringValue] == nil) ?
+        @"NSApplication" : [appCustomClass stringValue];
+      NSXMLElement *customObject = nil;
```
As I previously explained you could extract this duplicated code into a helper method:

- (void) addStandardObjects: ( NSXMLElement *)objects classString: (NSString *) customClassString
 connections: (NSXMLNode *)appCons
{
NSXMLElement *customObject = nil;

       customObject = 
         [self createCustomObjectWithId: @"-3"
                              userLabel: @"Application"
                            customClass: @"NSObject"];
       [objects insertChild: customObject
                  atIndex: 0];
       customObject = 
         [self createCustomObjectWithId: @"-1"
                              userLabel: @"First Responder"
                            customClass: @"FirstResponder"];
       [objects insertChild: customObject
                  atIndex: 0];
       customObject =
         [self createCustomObjectWithId: @"-2"
                              userLabel: @"File's Owner"
                            customClass: customClassString];
       if (appCons != nil)
         {
           [customObject addChild: appCons];
         }
       [objects insertChild: customObject
                  atIndex: 0]; 
}

Calling that method from both branches would remove the code duplication.

gcasa replied:
OMG, this is endless

gcasa replied:
Done.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,875 @@
+        }
+      else
+        {
+          NSString *destId = [_storyboardSegue destinationController];
```
Only in this case is the destination controller ever changed. It would be clearer if you checked the destination controller for being of class NSString and handled only that case:

if ([[_storyboardSegue destinationController] isKindOfClass: [NSViewController class]])
{
           NSString *destId = [_storyboardSegue destinationController];
destCon = [_storyboard instantiateControllerWithIdentifier: destId]; 
 // replace with actual controller...
[_storyboardSegue _setDestinationController: destCon]; 
         }

gcasa replied:
Ah, quite true.

gcasa replied:
Optimized this.

gcasa commented:
The code is much tighter and cleaner now.   Thanks, Fred, for all of your help.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,835 @@
+- (NSString *) controllerIdWithDocument: (NSXMLDocument *)document
+{
+  NSString *controllerId = nil;
+  NSArray *windowControllers = [self findSubclassesOf: [NSWindowController class]
```
Is this search for subclasses of windowController and viewController really needed? I look through a few projects on GitHub but couldn't find any using this. I know this does not prove anything. There is a customClass attribute on the XML node, so I would not expect to see a different tag for the subclasses. Do you have any example where this gets used? 
Removing this code would again simply the process here. 
I also only found at most one controller per scene. Do you have any more information on that?

fredkiefer replied:
I finally found one such subclass: navigationController, which seems to be important for UIKit. Currently we don't support that and even then it would be more useful to just add that string here directly.

gcasa replied:
Yes, because there may be subclasses that need also to be used.  NSTabViewController and NSPageController... any subclass of NSViewController which might be added in the future.  I did this to future-proof the code.

gcasa replied:
https://developer.apple.com/documentation/appkit/nstabviewcontroller?language=objc

https://developer.apple.com/documentation/appkit/nssplitviewcontroller?language=objc

gcasa replied:
For instance NSSplitViewController comes up as follows in the storyboard...

`+        <scene sceneID="Qzn-Ms-zfD">
+            <objects>
+                <splitViewController id="Eua-kV-LFV" sceneMemberID="viewController">
+                    <splitViewItems>
+                        <splitViewItem id="seP-J7-ftb"/>
+                        <splitViewItem id="IZ4-Aw-WYN"/>
+                    </splitViewItems>
+                    <splitView key="splitView" dividerStyle="thin" vertical="YE:
`

gcasa replied:
The three controllers mentioned are next on my list to be implemented, so this code will be further tested once that's done.

gcasa commented:
Fred, please take a look.  I believe all expectations have been met.

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,280 @@
+            {
+              NSStoryboardSeguePerformAction *ssa = (NSStoryboardSeguePerformAction *)o;
+              NSMapTable *mapTable = [_transform segueMapForIdentifier: identifier];
+              NSStoryboardSegue *ss = [mapTable objectForKey: [ssa identifier]];
```
You could use the variable segueMap here which has the same value.

gcasa replied:
Done

fredkiefer commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,280 @@
+      END_FOR_IN(placeholders);
+
+      // perform segues after all is initialized.
+      FOR_IN(id, o, seguesToPerform)
```
You could save you the type cast here (and in some other places) by already using the correct type in the loop:

 FOR_IN(NSStoryboardSeguePerformAction *, o, seguesToPerform)

gcasa replied:
Done

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+  
+  FOR_IN (NSXMLElement*, connObj, connectionsArray)
+    {
+      NSXMLElement *connections = (NSXMLElement *)connObj;
```
This line is pretty useless. Just use the name "connections" directly in the loop.

gcasa replied:
Done

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+      NSXMLElement *connections = (NSXMLElement *)connObj;
+      NSArray *children = [connections children]; // there should be only one per set.
+      
+      FOR_IN (id, obj, children)
```
Here again you should be using the type NSXMLElement*

gcasa replied:
Done

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+            {
+              // get the information from the segue.
+              id segue_parent_parent = [[obj parent] parent];
+              id segue_parent = [obj parent];
```
The variable "segue_parent" is always the same as "connections" and "segue_parent_parent" is also the same during one execution of this method.

gcasa replied:
segue_parent_parent should be the object which the segue it belongs to (aside from the connections).  Usually a controller.  This is why there is a check below.

gcasa replied:
I set segue_parent to connections.  I also renamed the "segue_parent_parent" to connections_parent.   The issue here is that it could be "application"... if it is then we want to take certain actions.  I believe this is resolved.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+      // Set initial view controller...
+      ASSIGN(_initialViewControllerId, [[docNode attributeForName: @"initialViewController"] stringValue]);             
+      FOR_IN(NSXMLElement*, e, array) 
+        {
```
Just a reminder as you probably already know this: You don't need the "{” after a FOR_IN statement. There is already one automatically added. It doesn't do any harm, you may always add braces in your code as you like.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+                         sceneId: sceneId];
+              
+              // fix other custom objects
+              document = [[NSXMLDocument alloc] initWithRootElement: doc]; // put it into the document, so we can use Xpath.
```
The comment here is a bit surprising as you could use nodesForXPath: on an NSXMLNode already.

gcasa replied:
Removed.

gcasa commented:
> I think there is still a lot to improve on this pull request and I am pretty sure that won't happen after I approve it but let's test this assumption. I will approve now and check back after a month to see what happened to the code.

I really wonder why this deliberately hurtful and, frankly, deeply unprofessional comment is necessary as it is a deliberate and, personal, attack on me, as usual.  I told you earlier that there are controllers that will be dependent on this code which, by definition, means it will be worked on.  Let it be known that we worked together on this for almost 2 weeks and I took ever one of your changes and suggestions into account.   You have an amazing amount of contempt for me, Fred, given the volume and quality of work that I have present in GNUstep.

If you feel like there is a lot to improve on this pull request I will hold off merging until tomorrow morning my time so that you can have ample opportunity to suggest changes.  If I don't receive any suggestions up until that point, then I will consider the matter closed.

gcasa commented:
Holding off merging due to Fred's asshattery.

fredkiefer commented:
> Holding off merging EXPLICITLY due to Fred's unnecessary ASSHATTERY.

As for "deliberately hurtful and, frankly, deeply unprofessional comment(s)" you have surely out done me. I can only suggest that you find some programmer that you respect and ask him/her to go over the history of this pull request and check whether he/she shares your view.

As for suggesting changes, I already added a few even before approval. Looks like you missed those.

To give you a bit of background why I am so sceptical about merging working code of questionable quality into GNUstep gui, I would like to remind you about our text system. This was written about fifteen years ago by Alexander Malmberg and it is mostly working. But any attempt to change even minor things in this code causes trouble every time as the code is unmaintainable. Yes, I might be slowing down the acceptance of new features in GNUstep gui a bit but the overall maintainability of that code has improved over the years (again slowly). If this by itself has no value for you then we just disagree here and you will have to live with my opinion.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+}
+@end
+
+@implementation NSStoryboardSeguePerformAction
```
I am having trouble understanding the concept of this class. There is one set of attributes that get decode from the XIB, target, action, selector. None of these ever gets used. There is another set storyboard, storyboardsegue that get used in the code. The two variables sender and identifier are somewhat in between as they actually get used. Could you please provide me of a real live example of this class in a storyboard?

fredkiefer replied:
OK, I think I understand it a bit better now. This class is no Apple class. The corresponding XML elements get created on the fly in createStoryboardProxyElementWithSelector:... that is why I could not find any real live example of that class.
It is here to represent the segue reference that was in the connection. Wouldn't it be better to actually decode this reference in GSXib5KeyedUnarchiver and treat it as another kind of connection there, like action, outlet and binding?

fredkiefer replied:
Why did you mark this as resolved?

gcasa replied:
I thought I had put an answer here, but it wasn't sent.   The reason I didn't do that, though I did consider it, was because of NSStoryboardSegue: 
1. Doesn't conform to NSCopying/NSCoding
2. Doesn't have 'target' and 'action' methods
3. This would have meant either creating a special case in NSStoryboardSegue or creating a special case in GSXib5KeyedUnarchiver.
4. Using a proxy allows for a place that the resolution of the controllers happens separately from the controller or segue code.

fredkiefer replied:
You would not create an NSStoryboardSegue directly there, so 1 is invalid. What we would need to add to GSXib5KeyedUnarchiver is "segue" as another connection and a mapping from "segue" to an NSConnection subclass, which could be something like NSStoryboardSeguePerformAction but called GSStoryboardSegueConnection, which isn't much of special casing, so 3 is invalid as well. As for 2, neither does an outlet or binding connection has an action. You might have a point in 4 as the instantiation of the segue seems to happen at a different point in time, but I am really not sure here. That code is just confusing.

gcasa replied:
Is this something we are talking about doing prior to merge?  This seems like a serious rewrite.

gcasa replied:
Actually, the reason this can't be done that way is that the destination of this object should not be instantiated the connection is invoked.  Under normal circumstances (like with an action, outlet, binding) the  connection is made via the establishConnection method.   In the case of bindings by making a "private" subclass and invoking code before and after the accessor methods are called.   I feel better having the NSStoryboardPerformAction (though I could probably rename it) since it ensures number 4 above.

fredkiefer replied:
As I wrote I am not sure whether that solution would be feasible and I must admit I did not fully understand your answer. It would definitely be worthwhile to test that approach though. At least we would need an example where this approach would not work.
Anyway, I am away for the weekend feel free to do what ever you want with this pull request.

fredkiefer replied:
Just to explain the proposed approach a bit more. I think when instantiating the GSStoryboardSegueConnection (or what ever we call it) this will just result in the corresponding segue being created and stores in the controllers map (very similar to the current code). But that map would only exist on the controller. The question is still where and when doAction: (or what ever replaces it) would be called. Perhaps you could explain where this currently happens?

gcasa replied:
okay, to break it down...

A normal connection object is created when "establishConnection" is called.  In all other connection classes this method is called unconditionally and sets the appropriate attributes to make the connection possible.  target/action in the case of an action and it associates the setter/getter (or instance var) in the case of an outlet.   In the case of a segue the destination controller is not instantiated unless the "shouldPerform" method returns YES.   This cannot happen in a traditional connection.  Therefore we cannot model a segue as a traditional connection.  This approach occurred to me when I originally did this.  In order to delay the instantiation of the destination controller I reasoned that there should be an object which acts as a go-between.

The other issue is that I didn't want to change the XIB decoding too much and figured it would be better to create an object in the XML which would be decoded on the fly and which could conform to NSCopying/NSCoding where NSStoryboardSegue could not.  Your connection class would achieve that goal, but not the other one.

gcasa replied:
This approach merits further discussion and thought.

gcasa replied:
I am looking at the code and thinking about this tonight.  I will report my findings later on.  I don't know if it is possible to model this purely as a connection, but I don't want to dismiss the possibility just yet.

gcasa replied:
I think the current implementation is sufficient at this point.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+  [super dealloc];
+}
+
+- (IBAction) doAction: (id)sender
```
This parameter gets ignored in the method. It is better that way as the parameter will mostly be nil but what is the idea of having it at all and computing it in the first place?

gcasa replied:
I presume youre referring to doAction:   it includes sender only because that is convention for an action.   I believe when it is called the calling object is set.

gcasa replied:
This is done by convention, I'm going to leave this as is.

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+        }
+      [_storyboardSegue _setSourceController: _sender]; 
+      
+      if (_sender != nil &&
```
The call to prepareForSegue:sender: is missing here.

fredkiefer replied:
Not true, currently it is called from within performSegueWithIdentifier:sender:. I am rather not sure whether this is the expected way to implement this

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+  return controllerId;
+}
+
+- (void) processCustomElement: (NSXMLElement *)coel
```
I was sure that I put in a comment to this method, but now I am no longer able to find it. Sometimes GitHub can play tricks on you. I appologize if I add this before.

The whole method could be replaced by just extracting the id of the first responder early on and putting that id into the processChild:... method using it instead of "-1". That way the actual unarchiving would resolve all the reference for us. As far as I am aware we never use the fact that the first responder has the id "-1" in our code.

fredkiefer replied:
You could extract that id with just on xpath statement. Something like `//objects/customObject[@sceneMemberID ="firstResponder"]/@id`

gcasa replied:
Looking into doing this.

gcasa replied:
Removed.

fredkiefer commented on Source/GSStoryboardTransform.h:
```diff
@@ -0,0 +1,120 @@
+- (NSData *) dataForApplicationScene;
+- (NSMapTable *) segueMapForIdentifier: (NSString *)identifier;
+
+- (void) processSegues: (NSXMLDocument *)xmlIn
```
These two methods could now be private.

gcasa replied:
Done

fredkiefer commented on Source/GSStoryboardTransform.m:
```diff
@@ -0,0 +1,828 @@
+      NSArray *array = [xml nodesForXPath: @"//objects[1]"
+                                    error: NULL];
+      NSXMLElement *objects = [array objectAtIndex: 0]; // get the "objects" section
+      NSString *src = [self controllerIdWithDocument: xml];                            
```
Looking at the code in processStoryboard: I would say that the value of "identifier" is always the same as the one you get here for "src" except for the application case. If that difference is important it could be resolved easily here.

gcasa replied:
Eliminated the call.  It is, indeed, unnecessary.

gcasa commented on Source/NSStoryboard.m:
```diff
@@ -0,0 +1,280 @@
+              NSMapTable *mapTable = [_transform segueMapForIdentifier: identifier];
+              NSStoryboardSegue *ss = [mapTable objectForKey: [ssa identifier]];
+              
+              [ssa setSender: result]; // resolve controller here...
```
corrected.  it now instantiates the placeholder before it reaches that point.  The "result" variable should now have the correct value.

gcasa commented:
Please take a look at this again.  I have implemented or addressed all of the current concerns.

gcasa commented:
Merging.  Fred had given the ok earlier and I have addressed all subsequent concerns.
